File: src/codeheart_operating_kit/components.py
from __future__ import annotations

import shutil
from datetime import timedelta
from pathlib import Path
from typing import Any

from . import __version__
from .lockfile import format_time, utc_now, write_config, write_lock
from .manifest import iter_component_files, kit_root, load_profile, sha256_file


BEGIN_MARKER = "<!-- BEGIN CODEHEART OPERATING KIT MANAGED BLOCK -->"
END_MARKER = "<!-- END CODEHEART OPERATING KIT MANAGED BLOCK -->"
LOCAL_USER_GITIGNORE_LINES = [
    "# Codeheart Operating Kit local user layer",
    ".codeheart/user/preferences.yaml",
    ".codeheart/user/*.local.yaml",
    ".codeheart/user/feedback/",
]
LOCAL_MACHINE_GITIGNORE_LINES = [
    "# Codeheart Operating Kit local machine layer",
    ".codeheart/local/",
]
LOCAL_GITIGNORE_LINES = LOCAL_USER_GITIGNORE_LINES + LOCAL_MACHINE_GITIGNORE_LINES


def copy_managed_files(root: Path, profile_id: str = "standard") -> list[dict[str, Any]]:
    records: list[dict[str, Any]] = []
    for entry in iter_component_files(profile_id):
        if entry.get("ownership") != "managed":
            continue
        source = kit_root() / entry["source"]
        target = root / entry["target"]
        target.parent.mkdir(parents=True, exist_ok=True)
        shutil.copyfile(source, target)
        records.append(
            {
                "path": entry["target"],
                "component": entry["component"],
                "source": entry["source"],
                "ownership": "managed",
                "checksum_sha256": sha256_file(target),
            }
        )
    return records


def render_agents(root: Path) -> str:
    template = (kit_root() / "templates/agents/AGENTS.managed-block.md").read_text(encoding="utf-8")
    existing_path = root / "AGENTS.md"
    if not existing_path.exists():
        existing_path.write_text(template, encoding="utf-8")
        return "created"

    existing = existing_path.read_text(encoding="utf-8")
    if BEGIN_MARKER in existing and END_MARKER in existing:
        before, rest = existing.split(BEGIN_MARKER, 1)
        _old, after = rest.split(END_MARKER, 1)
        managed = template.split(BEGIN_MARKER, 1)[1].split(END_MARKER, 1)[0]
        existing_path.write_text(f"{before}{BEGIN_MARKER}{managed}{END_MARKER}{after}", encoding="utf-8")
        return "repaired-managed-block"

    existing_path.write_text(template + "\n\n# Existing Instructions Preserved\n\n" + existing, encoding="utf-8")
    return "added-managed-block"


def refresh_agents_managed_block(root: Path) -> str:
    template = (kit_root() / "templates/agents/AGENTS.managed-block.md").read_text(encoding="utf-8")
    existing_path = root / "AGENTS.md"
    if not existing_path.exists():
        return "missing"

    existing = existing_path.read_text(encoding="utf-8")
    if BEGIN_MARKER not in existing or END_MARKER not in existing:
        return "unchanged-no-managed-block"

    before, rest = existing.split(BEGIN_MARKER, 1)
    _old, after = rest.split(END_MARKER, 1)
    managed = template.split(BEGIN_MARKER, 1)[1].split(END_MARKER, 1)[0]
    existing_path.write_text(f"{before}{BEGIN_MARKER}{managed}{END_MARKER}{after}", encoding="utf-8")
    return "refreshed-managed-block"


def scaffold_consumer_files(root: Path, profile_id: str = "standard") -> list[dict[str, str]]:
    created: list[dict[str, str]] = []
    for entry in iter_component_files(profile_id):
        if entry.get("ownership") != "scaffold":
            continue
        source = kit_root() / entry["source"]
        target = root / entry["target"]
        if target.exists():
            continue
        target.parent.mkdir(parents=True, exist_ok=True)
        shutil.copyfile(source, target)
        created.append({"path": entry["target"], "ownership": "scaffold"})

    user_dir = root / ".codeheart/user"
    user_dir.mkdir(parents=True, exist_ok=True)
    user_readme = user_dir / "README.md"
    if not user_readme.exists():
        shutil.copyfile(kit_root() / "templates/user-layer/README.md", user_readme)
        created.append({"path": ".codeheart/user/README.md", "ownership": "local-user"})
    examples = user_dir / "examples"
    examples.mkdir(exist_ok=True)
    example_pref = examples / "preferences.yaml"
    if not example_pref.exists():
        shutil.copyfile(kit_root() / "templates/user-layer/example.preferences.yaml", example_pref)
        created.append({"path": ".codeheart/user/examples/preferences.yaml", "ownership": "local-user"})
    return created


def ensure_gitignore(root: Path) -> bool:
    path = root / ".gitignore"
    text = path.read_text(encoding="utf-8") if path.exists() else ""
    lines = text.splitlines()
    missing = [line for line in LOCAL_GITIGNORE_LINES if line not in lines]
    if not missing:
        return False
    additions = []
    if lines:
        additions.append("")
    additions.extend(missing)
    path.write_text("\n".join(lines + additions).lstrip("\n") + "\n", encoding="utf-8")
    return True


def write_default_state(root: Path, project_name: str, purpose: str | None, selected_folder: str) -> dict[str, Any]:
    now = utc_now()
    profile = load_profile("standard")
    managed_records = copy_managed_files(root)
    scaffold_records = scaffold_consumer_files(root)
    agents_status = render_agents(root)
    gitignore_changed = ensure_gitignore(root)

    generated_surfaces = [
        {"path": ".codeheart/kit/", "ownership": "managed"},
        {"path": ".codeheart/kit.lock.yaml", "ownership": "generated-surface"},
        {"path": ".codeheart/kit.config.yaml", "ownership": "generated-surface"},
        {"path": "AGENTS.md", "ownership": "template"},
    ] + scaffold_records

    lock = {
        "schema_version": 1,
        "kit_version": __version__,
        "selected_profile": "standard",
        "selected_components": profile["selected_components"],
        "release": {
            "asset_url": "local-source",
            "checksum_sha256": "0" * 64,
        },
        "managed_paths": managed_records,
        "generated_surfaces": generated_surfaces,
        "cli_repair": {
            "installed_cli_path": "codeheart-operating-kit",
            "repair_source_url": "local-source",
            "repair_checksum_sha256": "0" * 64,
        },
        "update_check": {
            "last_update_check_at": format_time(now),
            "next_update_check_due": format_time(now + timedelta(days=7)),
            "latest_seen_version": __version__,
            "update_status": "current",
        },
        "native_capabilities": {
            capability: {
                "status": "unknown",
                "checked_at": format_time(now),
                "profile_applicability": "standard",
                "command_result_category": "not-checked",
            }
            for capability in ["documents", "spreadsheets", "presentations", "browser", "pdf"]
        },
    }
    write_lock(root, lock)
    config = {
        "schema_version": 1,
        "selected_profile": "standard",
        "project_display_name": project_name,
        "selected_setup_folder": selected_folder,
        "local_consumer_layer": {
            "repo_docs_path": "docs/repo/",
            "agent_memory_path": "docs/agent-memory/",
            "user_layer_path": ".codeheart/user/",
            "local_machine_layer_path": ".codeheart/local/",
        },
        "component_settings": {},
    }
    if purpose:
        config["setup_purpose"] = purpose
    write_config(root, config)
    return {
        "managed_paths": managed_records,
        "generated_surfaces": generated_surfaces,
        "agents_status": agents_status,
        "gitignore_changed": gitignore_changed,
    }


def write_adoption_cleanup_report(root: Path, findings: list[str]) -> Path:
    report = root / ".codeheart/reports/adoption-cleanup-report.md"
    report.parent.mkdir(parents=True, exist_ok=True)
    body = [
        "Last updated: 2026-06-13T00:00:00Z (UTC)",
        "",
        "# Adoption Cleanup Report",
        "",
        "Existing project guidance was preserved. Review these overlapping surfaces before cleanup:",
        "",
    ]
    body.extend(f"- {item}" for item in findings)
    report.write_text("\n".join(body) + "\n", encoding="utf-8")
    return report

File: internal/components/components.go
package components

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/Codeheart-Digital-Solutions/Codeheart-Operating-Kit/internal/capabilities"
	"github.com/Codeheart-Digital-Solutions/Codeheart-Operating-Kit/internal/hash"
	"github.com/Codeheart-Digital-Solutions/Codeheart-Operating-Kit/internal/kitfs"
	"github.com/Codeheart-Digital-Solutions/Codeheart-Operating-Kit/internal/lockfile"
	"github.com/Codeheart-Digital-Solutions/Codeheart-Operating-Kit/internal/manifest"
	"github.com/Codeheart-Digital-Solutions/Codeheart-Operating-Kit/internal/version"
)

const (
	BeginMarker = "<!-- BEGIN CODEHEART OPERATING KIT MANAGED BLOCK -->"
	EndMarker   = "<!-- END CODEHEART OPERATING KIT MANAGED BLOCK -->"
)

var localGitignoreLines = []string{
	"# Codeheart Operating Kit local user layer",
	".codeheart/user/preferences.yaml",
	".codeheart/user/*.local.yaml",
	".codeheart/user/feedback/",
	"# Codeheart Operating Kit local machine layer",
	".codeheart/local/",
}

type DefaultState struct {
	ManagedPaths      []map[string]any
	GeneratedSurfaces []map[string]any
	AgentsStatus      string
	GitignoreChanged  bool
}

func CopyManagedFiles(root string, profileID string) ([]map[string]any, error) {
	records := []map[string]any{}
	files, err := manifest.IterComponentFiles(profileID)
	if err != nil {
		return nil, err
	}
	for _, entry := range files {
		if entry.Ownership != "managed" {
			continue
		}
		targetPath := filepath.Join(root, filepath.FromSlash(entry.Target))
		if err := copyResource(entry.Source, targetPath); err != nil {
			return nil, err
		}
		checksum, err := hash.FileSHA256(targetPath)
		if err != nil {
			return nil, err
		}
		records = append(records, map[string]any{
			"path":            entry.Target,
			"component":       entry.Component,
			"source":          entry.Source,
			"ownership":       "managed",
			"checksum_sha256": checksum,
		})
	}
	return records, nil
}

func RenderAgents(root string) (string, error) {
	template, err := kitfs.ReadText("templates/agents/AGENTS.managed-block.md")
	if err != nil {
		return "", err
	}
	path := filepath.Join(root, "AGENTS.md")
	if _, err := os.Stat(path); os.IsNotExist(err) {
		return "created", os.WriteFile(path, []byte(template), 0o644)
	}
	existing, err := os.ReadFile(path)
	if err != nil {
		return "", err
	}
	text := string(existing)
	if strings.Contains(text, BeginMarker) && strings.Contains(text, EndMarker) {
		refreshed, err := replaceManagedBlock(text, template)
		if err != nil {
			return "", err
		}
		return "repaired-managed-block", os.WriteFile(path, []byte(refreshed), 0o644)
	}
	body := template + "\n\n# Existing Instructions Preserved\n\n" + text
	return "added-managed-block", os.WriteFile(path, []byte(body), 0o644)
}

func RefreshAgentsManagedBlock(root string) (string, error) {
	template, err := kitfs.ReadText("templates/agents/AGENTS.managed-block.md")
	if err != nil {
		return "", err
	}
	path := filepath.Join(root, "AGENTS.md")
	existing, err := os.ReadFile(path)
	if os.IsNotExist(err) {
		return "missing", nil
	}
	if err != nil {
		return "", err
	}
	text := string(existing)
	if !strings.Contains(text, BeginMarker) || !strings.Contains(text, EndMarker) {
		return "unchanged-no-managed-block", nil
	}
	refreshed, err := replaceManagedBlock(text, template)
	if err != nil {
		return "", err
	}
	return "refreshed-managed-block", os.WriteFile(path, []byte(refreshed), 0o644)
}

func ScaffoldConsumerFiles(root string, profileID string) ([]map[string]any, error) {
	created := []map[string]any{}
	files, err := manifest.IterComponentFiles(profileID)
	if err != nil {
		return nil, err
	}
	for _, entry := range files {
		if entry.Ownership != "scaffold" {
			continue
		}
		targetPath := filepath.Join(root, filepath.FromSlash(entry.Target))
		if exists(targetPath) {
			continue
		}
		if err := copyResource(entry.Source, targetPath); err != nil {
			return nil, err
		}
		created = append(created, map[string]any{"path": entry.Target, "ownership": "scaffold"})
	}

	userDir := filepath.Join(root, ".codeheart", "user")
	if err := os.MkdirAll(userDir, 0o755); err != nil {
		return nil, err
	}
	userReadme := filepath.Join(userDir, "README.md")
	if !exists(userReadme) {
		if err := copyResource("templates/user-layer/README.md", userReadme); err != nil {
			return nil, err
		}
		created = append(created, map[string]any{"path": ".codeheart/user/README.md", "ownership": "local-user"})
	}
	examplesDir := filepath.Join(userDir, "examples")
	if err := os.MkdirAll(examplesDir, 0o755); err != nil {
		return nil, err
	}
	examplePref := filepath.Join(examplesDir, "preferences.yaml")
	if !exists(examplePref) {
		if err := copyResource("templates/user-layer/example.preferences.yaml", examplePref); err != nil {
			return nil, err
		}
		created = append(created, map[string]any{"path": ".codeheart/user/examples/preferences.yaml", "ownership": "local-user"})
	}
	return created, nil
}

func EnsureGitignore(root string) (bool, error) {
	path := filepath.Join(root, ".gitignore")
	var lines []string
	if data, err := os.ReadFile(path); err == nil {
		lines = strings.Split(strings.ReplaceAll(string(data), "\r\n", "\n"), "\n")
		if len(lines) > 0 && lines[len(lines)-1] == "" {
			lines = lines[:len(lines)-1]
		}
	} else if !os.IsNotExist(err) {
		return false, err
	}
	present := map[string]bool{}
	for _, line := range lines {
		present[line] = true
	}
	missing := []string{}
	for _, line := range localGitignoreLines {
		if !present[line] {
			missing = append(missing, line)
		}
	}
	if len(missing) == 0 {
		return false, nil
	}
	additions := []string{}
	if len(lines) > 0 {
		additions = append(additions, "")
	}
	additions = append(additions, missing...)
	body := strings.Join(append(lines, additions...), "\n")
	body = strings.TrimLeft(body, "\n") + "\n"
	return true, os.WriteFile(path, []byte(body), 0o644)
}

func WriteDefaultState(root string, projectName string, purpose string, selectedFolder string, now time.Time) (DefaultState, error) {
	if err := os.MkdirAll(root, 0o755); err != nil {
		return DefaultState{}, err
	}
	profile, err := manifest.LoadProfile("standard")
	if err != nil {
		return DefaultState{}, err
	}
	managedRecords, err := CopyManagedFiles(root, "standard")
	if err != nil {
		return DefaultState{}, err
	}
	scaffoldRecords, err := ScaffoldConsumerFiles(root, "standard")
	if err != nil {
		return DefaultState{}, err
	}
	agentsStatus, err := RenderAgents(root)
	if err != nil {
		return DefaultState{}, err
	}
	gitignoreChanged, err := EnsureGitignore(root)
	if err != nil {
		return DefaultState{}, err
	}

	generatedSurfaces := []map[string]any{
		{"path": ".codeheart/kit/", "ownership": "managed"},
		{"path": ".codeheart/kit.lock.yaml", "ownership": "generated-surface"},
		{"path": ".codeheart/kit.config.yaml", "ownership": "generated-surface"},
		{"path": "AGENTS.md", "ownership": "template"},
	}
	generatedSurfaces = append(generatedSurfaces, scaffoldRecords...)

	now = now.UTC().Truncate(time.Second)
	nextCheck := now.Add(7 * 24 * time.Hour)
	lock := map[string]any{
		"schema_version":      1,
		"kit_version":         version.Version,
		"selected_profile":    "standard",
		"selected_components": stringSliceAsAny(profile.SelectedComponents),
		"release":             map[string]any{"asset_url": "local-source", "checksum_sha256": strings.Repeat("0", 64)},
		"managed_paths":       mapsAsAny(managedRecords),
		"generated_surfaces":  mapsAsAny(generatedSurfaces),
		"cli_repair":          map[string]any{"installed_cli_path": "codeheart-operating-kit", "repair_source_url": "local-source", "repair_checksum_sha256": strings.Repeat("0", 64)},
		"update_check":        map[string]any{"last_update_check_at": lockfile.FormatTime(now), "next_update_check_due": lockfile.FormatTime(nextCheck), "latest_seen_version": version.Version, "update_status": "current"},
		"native_capabilities": capabilities.UnknownNativeCapabilityState(now),
	}
	if err := lockfile.WriteLock(root, lock); err != nil {
		return DefaultState{}, err
	}

	config := map[string]any{
		"schema_version":        1,
		"selected_profile":      "standard",
		"project_display_name":  projectName,
		"selected_setup_folder": selectedFolder,
		"local_consumer_layer": map[string]any{
			"repo_docs_path":           "docs/repo/",
			"agent_memory_path":        "docs/agent-memory/",
			"user_layer_path":          ".codeheart/user/",
			"local_machine_layer_path": ".codeheart/local/",
		},
		"component_settings": map[string]any{},
	}
	if purpose != "" {
		config["setup_purpose"] = purpose
	}
	if err := lockfile.WriteConfig(root, config); err != nil {
		return DefaultState{}, err
	}

	return DefaultState{
		ManagedPaths:      managedRecords,
		GeneratedSurfaces: generatedSurfaces,
		AgentsStatus:      agentsStatus,
		GitignoreChanged:  gitignoreChanged,
	}, nil
}

func WriteAdoptionCleanupReport(root string, findings []string) (string, error) {
	path := filepath.Join(root, ".codeheart", "reports", "adoption-cleanup-report.md")
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return "", err
	}
	lines := []string{
		"Last updated: 2026-06-13T00:00:00Z (UTC)",
		"",
		"# Adoption Cleanup Report",
		"",
		"Existing project guidance was preserved. Review these overlapping surfaces before cleanup:",
		"",
	}
	for _, finding := range findings {
		lines = append(lines, "- "+finding)
	}
	if err := os.WriteFile(path, []byte(strings.Join(lines, "\n")+"\n"), 0o644); err != nil {
		return "", err
	}
	return path, nil
}

func copyResource(source string, targetPath string) error {
	data, err := kitfs.ReadFile(source)
	if err != nil {
		return err
	}
	if err := os.MkdirAll(filepath.Dir(targetPath), 0o755); err != nil {
		return err
	}
	return os.WriteFile(targetPath, data, 0o644)
}

func replaceManagedBlock(existing string, template string) (string, error) {
	before, rest, ok := strings.Cut(existing, BeginMarker)
	if !ok {
		return "", fmt.Errorf("existing text has no begin marker")
	}
	_, after, ok := strings.Cut(rest, EndMarker)
	if !ok {
		return "", fmt.Errorf("existing text has no end marker")
	}
	_, managedRest, ok := strings.Cut(template, BeginMarker)
	if !ok {
		return "", fmt.Errorf("template has no begin marker")
	}
	managed, _, ok := strings.Cut(managedRest, EndMarker)
	if !ok {
		return "", fmt.Errorf("template has no end marker")
	}
	return before + BeginMarker + managed + EndMarker + after, nil
}

func exists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}

func mapsAsAny(values []map[string]any) []any {
	result := make([]any, len(values))
	for index, value := range values {
		result[index] = value
	}
	return result
}

func stringSliceAsAny(values []string) []any {
	result := make([]any, len(values))
	for index, value := range values {
		result[index] = value
	}
	return result
}
